### PR TITLE
Add device affinity to sharded tensors

### PR DESCRIPTION
**Pull request by rsuderman**
This adds some default device affinities to loaded sharded tensors. This still needs to be updated for a configurable allocation but enables our current models to run on per-tensor parallelism.

### CODE REVIEW

File: sharktank/sharktank/types/tensors.py [hunk 1]
```diff
@@ -1135,10 +1138,21 @@ def create(
     ) -> "InferenceTensor":
         shard_count = int(extra_properties["shard_count"])
         try:
-            ts = raw_tensors[""]
+            # We have to do this to avoid exporting as part of the `mlir` blob:
```
Comment (sogartar) [inline sharktank/sharktank/types/tensors.py:1141]: Maybe this will fix what I have been observing that some constants appear in the MLIR.

**Review by sogartar — APPROVED** [review]

File: sharktank/sharktank/ops/sharded_impls.py [hunk 2]
```diff
@@ -41,7 +41,7 @@ def all_gather_split(
     shards = [
         cat(
             [
-                transfer_to_logical_device(shard, i) if j != i else shard
+                transfer_to_logical_device(shard, i)
```
Comment (sogartar) [inline sharktank/sharktank/ops/sharded_impls.py:44]: Why did you remove this optimization?

File: sharktank/sharktank/types/tensors.py (hunk 1 above)
Comment (rsuderman) [reply sharktank/sharktank/types/tensors.py:1141]: Yes. The copying semantic drops the externalization for flow tensor transfer. I think we should ultimately drop the transfer in the ReplicateTensor construction for a `torch.Tensor`. I honestly wish we could drop it altogether but materialized constants are going to be an issue we need to track down.

File: sharktank/sharktank/ops/sharded_impls.py (hunk 2 above)
Comment (rsuderman) [reply sharktank/sharktank/ops/sharded_impls.py:44]: It causes compilation issues with the model sadly. It happens during some fusion passes so its likely a compiler bug but we can reintroduce it later on.
